# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: xpcom/build/Omnijar.cpp
# Commit: abcb98a505f6
# Full Hash: abcb98a505f6e9ce6c30051f7c80a39c78fee9cd
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-08-09 21:39:40
# Regressor Bug: 1627075
# File Overlap Count: 1
# Description:
#   Bug 1657205 - Fallibly allocate for StartupCache from Omnijar r=froydnj
#   
#   This changes the StartupCache::PutBuffer call from Omnijar to use a
#   fallibly allocated buffer, to reduce OOM crashes. We can/should broaden the
#   scope of this, but this is a simple initial change to mitigate the OOM
# ==============================================================================

diff -r baf18578315e -r abcb98a505f6 xpcom/build/Omnijar.cpp
--- a/xpcom/build/Omnijar.cpp	Sat Aug 08 15:47:04 2020 +0000
+++ b/xpcom/build/Omnijar.cpp	Fri Aug 07 20:05:59 2020 +0000
@@ -14,6 +14,7 @@
 #include "nsNetUtil.h"
 #include "mozilla/scache/StartupCache.h"
 #include "mozilla/MmapFaultHandler.h"
+#include "mozilla/UniquePtrExtensions.h"
 
 namespace mozilla {
 
@@ -438,12 +439,14 @@
   }
 
   auto* cache = scache::StartupCache::GetSingleton();
-  auto dataCopy = MakeUnique<char[]>(aSize);
+  auto dataCopy = MakeUniqueFallible<char[]>(aSize);
 
-  MMAP_FAULT_HANDLER_BEGIN_BUFFER(aBuffer, aSize)
-  memcpy(dataCopy.get(), aBuffer, aSize);
-  MMAP_FAULT_HANDLER_CATCH()
-  Unused << cache->PutBuffer(aCacheKey.get(), std::move(dataCopy), aSize);
+  if (dataCopy) {
+    MMAP_FAULT_HANDLER_BEGIN_BUFFER(aBuffer, aSize)
+    memcpy(dataCopy.get(), aBuffer, aSize);
+    MMAP_FAULT_HANDLER_CATCH()
+    Unused << cache->PutBuffer(aCacheKey.get(), std::move(dataCopy), aSize);
+  }
 }
 
 void CacheAwareZipReader::PushSuspendStartupCacheWrites() {
